backend/corpora/dataset_processing/upload.py

```diff
@@ -0,0 +1,139 @@
+        self.stop_updater: threading.Event = threading.Event()  # Stops the update_progress thread
+        self.stop_uploader: threading.Event = threading.Event()  # Stops the uploader threads
+        self.error: queue.Queue = queue.Queue(maxsize=1)  # Track errors
+
+    def progress(self):
+        with self.progress_lock:
+            return self._progress / self.file_size
+
+    def update(self, progress):
+        with self.progress_lock:
+            self._progress += progress
+
+
+def uploader(url: str, local_path: str, tracker: ProgressTracker, chunk_size: int):
+    """
+    Upload the file pointed at by the URL to the local path.
+
+    :param url: The URL of the file to be uploaded.
+    :param local_path: The local name of the file be uploaded
+    :param tracker: Tracks information about the progress of the upload.
+    :return:
+    """
+    try:
+        with requests.get(url, stream=True) as resp:
+            resp.raise_for_status()
+            with open(local_path, "wb") as fp:
+                for chunk in resp.iter_content(chunk_size=chunk_size):
+                    if tracker.stop_uploader.isSet():
+                        logger.debug("Upload ended early!")
+                        return
+                    elif chunk:
+                        fp.write(chunk)
+                        chunk_size = len(chunk)
+                        tracker.update(chunk_size)
+                        logger.debug(f"chunk size: {chunk_size}")
+    except requests.HTTPError as ex:
+        tracker.error.put(ex)
+    finally:
+        tracker.stop_updater.set()
+
+
+def processing_status_updater(uuid: str, updates: dict):
+    with db_session_manager() as manager:
+        manager.session.query(DbDatasetProcessingStatus).filter(DbDatasetProcessingStatus.id == uuid).update(updates)
+        manager.session.commit()
+
+
+def updater(upload_uuid: str, tracker: ProgressTracker, frequency: float = 3):
+    """
+    Update the progress of an upload to the database using the tracker.
+
+    :param upload_uuid: The uuid of the upload_progress row.
+    :param tracker: Tracks information about the progress of the upload.
+    :param frequency: The frequency in which the database is updated
+    :return:
+    """
+
+    def _update():
+        progress = tracker.progress()
+        if progress > 1:
+            tracker.stop_uploader.set()
+            processing_status = {
+                DbDatasetProcessingStatus.upload_progress: progress,
+                DbDatasetProcessingStatus.upload_message: "The file size, does not match the size of the upload.",
+                DbDatasetProcessingStatus.upload_status: UploadStatus.FAILED,
+            }
+        elif progress == 1 and tracker.stop_updater.isSet():
+            processing_status = {
+                DbDatasetProcessingStatus.upload_progress: progress,
+                DbDatasetProcessingStatus.upload_status: UploadStatus.UPLOADED,
+            }
+        else:
```

Comment:
I think we need to handle the case where the upload completes, but the local file it too small. I think currently is just stays UPLOADING forever.

Reply:
added